### PR TITLE
Make per-loop display clearing optional

### PR DESCRIPTION
**Pull request by timn**
# What does this implement/fix? 

This change adds a new optional flag "auto_clear" that can be used to control the clearing behavior of DisplayBuffer. If unset defaults to enabled auto clearing, the current behavior and thus backward compatible.

## Types of changes

- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Configuration change (this will require users to update their yaml configuration files to keep working)

**Pull request in [esphome-docs](https://github.com/esphome/esphome-docs) with documentation (if applicable):** esphome/esphome-docs#1096
  
# Test Environment

- [x] ESP32
- [ ] ESP8266
- [ ] Windows
- [ ] Mac OS
- [x] Linux

## Example entry for `config.yaml`:

```yaml
globals:
  - id: state
    type: bool
    restore_value: no
    initial_value: "false"
  - id: state_processed
    type: bool
    restore_value: no
    initial_value: "false"

switch:
  - platform: template
    name: "State"
    id: state_switch
    lambda: |-
      return id(state);
    turn_on_action:
      - globals.set:
          id: state
          value: "true"
      - globals.set:
          id: state_processed
          value: "false"
    turn_off_action:
      - globals.set:
          id: state
          value: "false"
      - globals.set:
          id: state_processed
          value: "false"

display:
  - platform: ili9341
    # ...
    auto_clear: false
    lambda: |-
      if (!id(state_processed)) {
        it.fill(COLOR_WHITE);
        if (id(state)) {
          it.image(80, 20, id(image1));
        } else {
          it.image(80, 20, id(image2));
        }
        id(state_processed) = true;
      }
```

# Explain your changes

Currently, in each loop during DisplayBuffer::update_() the display is cleared by calling DisplayBuffer::clear().

This prevents more efficient display usages that do not render the screen in each loop, but only if necessary. This can be helpful, for example, if images are rendered. This would cause the loop time to be exceeded frequently.

## Checklist:
  - [x] The code change is tested and works locally.
  - [x] Tests have been added to verify that the new code works (under `tests/` folder).
  
If user exposed functionality or configuration variables are added/changed:
  - [x] Documentation added/updated in [esphome-docs](https://github.com/esphome/esphome-docs).

### CODE REVIEW

**Comment by SenexCrenshaw** [comment]
Check out 
https://github.com/esphome/esphome/pull/1386
This will eventually make its way in. It is a complete overhaul of the buffer system including this piece.

File: tests/test1.yaml [hunk 1]
```diff
@@ -1904,6 +1904,37 @@ display:
...
+    cs_pin: GPIO5
+    dc_pin: GPIO4
+    reset_pin: GPIO22
+    led_pin:
+      number: GPIO15
+      inverted: true
+    auto_clear: false
+    rotation: 90
+    lambda: |-
+      if (!id(glob_bool_processed)) {
```
Comment (SenexCrenshaw) [inline tests/test1.yaml:1928]: Decouple this from the tm1651 component . The test should be generic to the display component. 
``it.fill(COLOR_WHITE);`` is enough

File: esphome/const.py [hunk 2]
```diff
@@ -66,6 +66,7 @@
 CONF_AT = "at"
 CONF_ATTENUATION = "attenuation"
 CONF_AUTH = "auth"
+CONF_AUTO_CLEAR = "auto_clear"
```
Comment (SenexCrenshaw) [inline esphome/const.py:69]: Should this be called ``clear_on_loop`` ? Auto clear is vague but I would not care if it stays ``auto_clear``

**Review by SenexCrenshaw — CHANGES_REQUESTED** [review]
Looks good. This is a handy addition.

File: esphome/const.py (hunk 2 above)
Comment (OttoWinter) [reply esphome/const.py:69]: I do think `auto_clear` is better, `on_loop` is probably something most people are not familiar with.

I'd add `_enabled` though, like `auto_clear_enabled.

**Pull request closed without merging**